Prepare to switch from dnf to rpm in our pipelines

teg opened the pull request:
These preliminary commits adds the needed low-level features to be able to use the rpm rather than the dnf stage. The switchover to change the produced pipelines is not yet done, so this can be merged independently of the required changes in osbuild.

msehnout commented on internal/weldr/api.go:
```diff
@@ -1757,7 +1757,14 @@ func (api *API) fetchPackageList() (rpmmd.PackageList, error) {
 }
 
-func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.PackageSpec, map[string]string, error) {
+func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, outputType, arch string, clean bool) ([]rpmmd.PackageSpec, []rpmmd.PackageSpec, map[string]string, error) {
```
I'd prefer if new code used the types from `internal/common` and variable names reflected what is written in the README file. So I'd suggest to replace `outputType` with `imageType`.

msehnout commented on internal/weldr/api.go:
```diff
@@ -1769,16 +1776,33 @@ func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.
 		}
 	}
+	excludeSpecs := []string{}
+	if outputType != "" {
```
How is it possible to reach this function without an image type? I thought depsolv without the set of base packages does not make any sense.

teg replied:
I decided to leave this in for now, and rather clean up all the naming in one commit/PR.

msehnout commented on internal/weldr/api.go:
```diff
@@ -1769,16 +1776,33 @@ func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.
+	if outputType != "" {
+		buildSpecs, err := api.distro.BuildPackages(arch)
+		if err != nil {
+			return nil, nil, nil, err
```
Do we have a reasonable way to tell the user what happened here? Was it possible to have such errors in Weldr?

teg replied:
This would be an programming error on our side. We should `panic()` or at most return `InternalServerError`.

msehnout commented on internal/osbuild/source.go:
```diff
@@ -0,0 +1,43 @@
+		case "org.osbuild.files":
+			source = new(FilesSource)
+		default:
+			return errors.New("unexpected suorce name" + name)
```
Just a small typo :)

msehnout replied:
Still there ;-)

jkozol commented on internal/distro/fedora30/distro.go:
```diff
@@ -436,16 +455,11 @@ func (r *Fedora30) Runner() string {
-	packages = append(packages, arch.BuildPackages...)
+	packages, err := r.BuildPackages(arch.Name)
+	if err != nil {
+		panic("impossibly invalid arch")
```
I agree that this is not a widely used phrase and "invalid arch" or "impossible and invalid arch" might be technically better but I think it explains the error fine and I personally enjoy the phrase "impossibly invalid arch".

teg replied:
I spent way too long trying to decide what to do with this feedback ;) In the end: yes, the phrase has a nice sound to it, but I'll make it grammatically correct.

teg commented:
Rebased on master, fixed according to the reviews and dropped the final commit that changes the output of osbuild-pipeline. This should be done after osbuild itself has been adjusted.

msehnout commented on internal/distro/test/distro.go:
```diff
@@ -0,0 +1,60 @@
+package test
```
Is this package used anywhere? I replaced all usages with the fedoratest packages if I remember correctly.

msehnout commented:
As other were fine with this PR already, I'm going to merge this now. Thanks @teg !
